fdbclient/include/fdbclient/DatabaseContext.h

```diff
@@ -730,7 +730,17 @@ class DatabaseContext : public ReferenceCounted<DatabaseContext>, public FastAll
 
 	WatchMap_t watchMap;
 
-	using WatchCounterMap_t = std::unordered_map<WatchMapKey, int32_t, WatchMapKeyHasher>;
+	// The reason of using a multiset of Versions as counter instead of a simpler integer counter is due to the
+	// possible race condition:
+	//
+	//    1. A watch to key A is set, the watchValueMap ACTOR, noted as X, starts waiting.
+	//    2. All watches are cleared due to connection string change.
+	//    3. The watch to key A is restarted with watchValueMap ACTOR Y.
+	//    4. X receives the cancel exception, and tries to dereference the counter. This causes Y gets cancelled.
+	//
+	// By introducing versions, this race condition is solved.
```

Comment:
The `multiset` is introduced to solve this issue. Since it accepts multiple copies of same value, I would suppose this issue will be solved without problem.

Reply:
Oh of course, that makes sense